### PR TITLE
[NativeAOT/ARM64] Generate frames compatible with Apple compact unwinding

### PR DESCRIPTION
**Pull request by filipnavara**
Contributes to #76371

The are two changes in the PR that work in tandem. The first is a JIT change for generating slightly different frame layout on NativeAOT/ARM64/Apple platforms (iOS, macOS and tvOS). The second change is ObjWriter code that recognizes the structurally compatible unwinding information and generates compact unwinding codes in the object files instead of verbose DWARF unwinding information.

For NativeAOT/ARM64/Apple ABI do the following:
1) Save callee registers in opposite order and in pairs.
2) Prefer saving FP/LR on the top of the frame. Heuristics are used to avoid worse code quality outside of prolog/epilog due to addressing range limits of the ARM64 instruction set.
3) Added optimization to `lvaFrameAddress` to rewrite FP-x references to SP+y when possible. This allows efficient addressing using positive indexes when FP points to the top of the frame. It mimics similar optimization on ARM32.

Each of these changes comes with some caveats:
1) Saving registers in pairs may lead to 16 bytes more used in the stack space. The ARM64 stack has to be 16 byte aligned. If the prolog previously saved odd number of integer callee saved registers and odd number of floating point callee saved registers, we'd now generate one more 16 byte stack slot. This seems to be rather rare occurrence. It does not affect the number of instructions used. Likewise, the changed order doesn't have any impact on code size.
2) Saving FP/LR at the top of the frame may cause one additional instruction in the prolog if any local or temporary variables are used. Additionally, addressing using negative offsets from FP is less efficient than addressing with positive offsets since the ARM64 instruction encoding can only address negative offsets with non-scaled encoding which significantly reduces the addressable range. This is mostly mitigated by the optimization in `lvaFrameAddress`. The additional increase in prolog size also causes a cascading effect where loop alignment and related code alignment (32-byte alignment for method start) significantly contribute to the code section size. In some cases the same alignment logic may also reduce the size.

As you can see, this is a bit of a trade-off and it's valid to ask if it's worth it.

Firstly, let me address the size question. The loop alignment seems to be the biggest contributor to the code size variation, and I filed issue https://github.com/dotnet/runtime/issues/107284 to investigate whether we can come up with a better defaults for Apple platforms. The code size changes are predominantly contained to the small fixed change in the prolog size and the additional alignment. When testing the prototype on MAUI apps the biggest culprit to increased code size was the code generated from XAML that generates methods with ton of local variables. The change in `lvaFrameAddress` practically eliminated any negative effect of the frame layout on this type of code. Without the change the regression was nearly 50% due to stack references needing an indirect load with a register and extra instruction. Outside of MAUI the biggest visible effect is on code from Regex source generator but in that case it's quite evenly split between size improvements and regressions. That suggests the regex code may be a good candidate for measuring the performance characteristics of the loop alignment. The code size regressions on few examples I tried amount to around 2% (incl. the alignment). The saved space in the DWARF unwinding section is hovering around 90% +- 3%. To put that into absolute numbers we are looking at savings around 0.7 Mb for `dotnet new maui` app and around 3 Mb for `System.Runtime.Tests` in the linked executables. The savings in the size of the unwinding information far outweigh any increase in the code size.

Secondly, part of the motivation is that the Apple linker is notoriously buggy with processing the DWARF unwinding data. The compact unwinding tables are used as an index to the DWARF data for anything that cannot be expressed using compact unwinding code directly. Due to the structure of the tables this limits the effective offset of DWARF info to 24 bits and places a hard limit on the DWARF unwinding info size. At least with some versions of the Apple linker, breaking this limit results in silent corruption and runtime failures.

<details>
<summary>Comparison between `main` and PR for System.Runtime.Tests in Release build</summary>

Compare raw size of linked binaries:
```
ls -l ./artifacts/bin/System.Runtime.Tests/Release/net9.0-unix/osx-arm64/publish/System.Runtime.Tests ../runtime-main/artifacts/bin/System.Runtime.Tests/Release/net9.0-unix/osx-arm64/publish/System.Runtime.Tests
-rwxr-xr-x  1 filipnavara  staff  48913112 Sep 12 22:26 ../runtime-main/artifacts/bin/System.Runtime.Tests/Release/net9.0-unix/osx-arm64/publish/System.Runtime.Tests
-rwxr-xr-x  1 filipnavara  staff  45827128 Sep 12 22:34 ./artifacts/bin/System.Runtime.Tests/Release/net9.0-unix/osx-arm64/publish/System.Runtime.Tests
```

Bloaty check of linked binaries:
```
bloaty -d symbols ./artifacts/bin/System.Runtime.Tests/Release/net9.0-unix/osx-arm64/publish/System.Runtime.Tests -- ../runtime-main/artifacts/bin/System.Runtime.Tests/Release/net9.0-unix/osx-arm64/publish/System.Runtime.Tests
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  +1.9%  +406Ki  +1.9%  +406Ki    [__TEXT,__managedcode]
   +55% +8.80Ki   +54% +8.80Ki    [__TEXT]
  +0.1% +2.51Ki  +0.1% +2.51Ki    [__DATA,.dotnet_eh_table]
  +0.1%    +272  +0.1%    +272    [__DATA,__data]
  -0.0%      -5  -0.0%      -5    [__TEXT,__cstring]
  -0.0%     -16  -0.0%     -16    [__TEXT,__const]
 -33.0% -2.78Ki -25.0% -2.78Ki    [__DATA]
  -3.6% -21.7Ki  -2.6% -16.0Ki    [__LINKEDIT]
 -20.5%  -437Ki -20.5%  -437Ki    [__TEXT,__unwind_info]
 -88.7% -2.90Mi -88.7% -2.90Mi    [__TEXT,__eh_frame]
  -6.3% -2.94Mi  -5.4% -2.94Mi    TOTAL
```

Bloaty check of object files:
```
bloaty ./artifacts/obj/System.Runtime.Tests/Release/net9.0-unix/osx-arm64/native/System.Runtime.Tests.o -- ../runtime-main/artifacts/obj/System.Runtime.Tests/Release/net9.0-unix/osx-arm64/native/System.Runtime.Tests.o 
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  +1.9%  +406Ki  +1.9%  +406Ki    ,__managedcode
  +0.1% +9.64Ki  +0.1% +9.64Ki    ,__debug_loc
  +0.1% +2.51Ki  +0.1% +2.51Ki    ,.dotnet_eh_table
  +0.0%    +348  +0.0%    +348    ,__debug_line
  +0.1%    +272  +0.1%    +272    ,__data
  +0.0%     +20  +0.0%     +20    ,__debug_info
  -0.1%      -1  -2.3%      -1    []
  -0.0%      -9  -0.0%      -9    ,__const
  -6.8% -1.76Mi  [ = ]       0    [Unmapped]
 -88.7% -2.90Mi -88.7% -2.90Mi    ,__eh_frame
  -2.4% -4.25Mi  -2.4% -2.49Mi    TOTAL
```

Bloaty check of object files (detailed):
```
bloaty -d symbols ./artifacts/obj/System.Runtime.Tests/Release/net9.0-unix/osx-arm64/native/System.Runtime.Tests.o -- ../runtime-main/artifacts/obj/System.Runtime.Tests/Release/net9.0-unix/osx-arm64/native/System.Runtime.Tests.o
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  +1.9%  +410Ki  +1.6%  +410Ki    [36252 Others]
  +0.1% +9.64Ki  +0.1% +9.64Ki    [,__debug_loc]
  [NEW] +7.19Ki  [NEW] +6.97Ki    _TestUtilities_Unicode_System_Text_RegularExpressions_Generated__RegexGenerator_g_F026929D4AD63EB6EDB749A9B02B133C084237B3AFB8777B3DE0107C755B91565__GetRegex_0_RunnerFactory_Runner__TryMatchAtCurrentPosition
  [NEW] +2.42Ki  [NEW] +2.20Ki    _S_P_Xml_System_Text_RegularExpressions_Generated__RegexGenerator_g_F417AD2970E27AC5D022778EC7081B94838CBCB072379596BB27CB8792C23ED76__EnsureArrayIndexRegex_5_RunnerFactory_Runner__TryMatchAtCurrentPosition
  [NEW] +2.02Ki  [NEW] +1.81Ki    _S_P_Xml_System_Text_RegularExpressions_Generated__RegexGenerator_g_F417AD2970E27AC5D022778EC7081B94838CBCB072379596BB27CB8792C23ED76__Regex1_3_RunnerFactory_Runner__TryMatchAtCurrentPosition
  [NEW] +1.54Ki  [NEW] +1.51Ki    _lsda0_TestUtilities_Unicode_System_Text_RegularExpressions_Generated__RegexGenerator_g_F026929D4AD63EB6EDB749A9B02B133C084237B3AFB8777B3DE0107C755B91565__GetRegex_0_RunnerFactory_Runner__TryMatchAtCurrentPosition
  [NEW] +1.44Ki  [NEW] +1.23Ki    _S_P_Xml_System_Text_RegularExpressions_Generated__RegexGenerator_g_F417AD2970E27AC5D022778EC7081B94838CBCB072379596BB27CB8792C23ED76__P0Regex_6_RunnerFactory_Runner__TryMatchAtCurrentPosition
  [NEW] +1.18Ki  [NEW]    +976    _System_Runtime_Tests_System_Text_RegularExpressions_Generated__RegexGenerator_g_FB4BB6619E624A9FE5F4E687DA0CF38E2970A0CF70BB059A398625E558ACC7DAC__IanaAbbreviationRegex_0_RunnerFactory_Runner__TryMatchAtCurrentPosition
  [NEW] +1.14Ki  [NEW]    +960    _S_P_Xml_System_Text_RegularExpressions_Generated__RegexGenerator_g_F417AD2970E27AC5D022778EC7081B94838CBCB072379596BB27CB8792C23ED76__Regex2_4_RunnerFactory_Runner__TryMatchAtCurrentPosition
  [NEW] +1.11Ki  [NEW]    +912    _S_P_Xml_System_Text_RegularExpressions_Generated__RegexGenerator_g_F417AD2970E27AC5D022778EC7081B94838CBCB072379596BB27CB8792C23ED76__UnknownNodeObjectEmptyRegex_8_RunnerFactory_Runner__TryMatchAtCurrentPosition
  [DEL]   -1017  [DEL]    -800    _S_P_Xml_System_Text_RegularExpressions_Generated__RegexGenerator_g_F24C2B164CF72F5E63C813A2B442D56F3F17BFFC74EAF7C2818EB7F6278C5183A__EncodeCharRegex_1_RunnerFactory_Runner__TryMatchAtCurrentPosition
  [DEL] -1.11Ki  [DEL]    -912    _S_P_Xml_System_Text_RegularExpressions_Generated__RegexGenerator_g_F24C2B164CF72F5E63C813A2B442D56F3F17BFFC74EAF7C2818EB7F6278C5183A__UnknownNodeObjectEmptyRegex_8_RunnerFactory_Runner__TryMatchAtCurrentPosition
  [DEL] -1.14Ki  [DEL]    -960    _S_P_Xml_System_Text_RegularExpressions_Generated__RegexGenerator_g_F24C2B164CF72F5E63C813A2B442D56F3F17BFFC74EAF7C2818EB7F6278C5183A__Regex2_4_RunnerFactory_Runner__TryMatchAtCurrentPosition
  [DEL] -1.17Ki  [DEL]    -960    _System_Runtime_Tests_System_Text_RegularExpressions_Generated__RegexGenerator_g_F23FFE14CED6C53CC123B603EF102D84787AD8CF9A59E83434F2BDF516814C2CC__IanaAbbreviationRegex_0_RunnerFactory_Runner__TryMatchAtCurrentPosition
  [DEL] -1.44Ki  [DEL] -1.23Ki    _S_P_Xml_System_Text_RegularExpressions_Generated__RegexGenerator_g_F24C2B164CF72F5E63C813A2B442D56F3F17BFFC74EAF7C2818EB7F6278C5183A__P0Regex_6_RunnerFactory_Runner__TryMatchAtCurrentPosition
  [DEL] -1.57Ki  [DEL] -1.54Ki    _lsda0_TestUtilities_Unicode_System_Text_RegularExpressions_Generated__RegexGenerator_g_F25E40F247ED975110B5E93851ECC9A2CB78C4DC17A73BF5F5EE6D76B142C41B6__GetRegex_0_RunnerFactory_Runner__TryMatchAtCurrentPosition
  [DEL] -2.00Ki  [DEL] -1.80Ki    _S_P_Xml_System_Text_RegularExpressions_Generated__RegexGenerator_g_F24C2B164CF72F5E63C813A2B442D56F3F17BFFC74EAF7C2818EB7F6278C5183A__Regex1_3_RunnerFactory_Runner__TryMatchAtCurrentPosition
  [DEL] -2.42Ki  [DEL] -2.20Ki    _S_P_Xml_System_Text_RegularExpressions_Generated__RegexGenerator_g_F24C2B164CF72F5E63C813A2B442D56F3F17BFFC74EAF7C2818EB7F6278C5183A__EnsureArrayIndexRegex_5_RunnerFactory_Runner__TryMatchAtCurrentPosition
  [DEL] -7.19Ki  [DEL] -6.97Ki    _TestUtilities_Unicode_System_Text_RegularExpressions_Generated__RegexGenerator_g_F25E40F247ED975110B5E93851ECC9A2CB78C4DC17A73BF5F5EE6D76B142C41B6__GetRegex_0_RunnerFactory_Runner__TryMatchAtCurrentPosition
  -6.8% -1.76Mi  [ = ]       0    [Unmapped]
 -88.7% -2.90Mi -88.7% -2.90Mi    lsection3
  -2.4% -4.25Mi  -2.4% -2.49Mi    TOTAL
```
</details>

### CODE REVIEW

**Comment by filipnavara** [comment]
The dominant part are the JIT changes, hence [area-CodeGen-coreclr](https://github.com/dotnet/runtime/labels/area-CodeGen-coreclr) is the right label.

cc @dotnet/ilc-contrib @VSadov

**Comment by marek-safar** [comment]
@jkotas would could help to move this PR forward?

**Review by jkotas — APPROVED** [review]
> @jkotas would could help to move this PR forward?

The change looks good to me. Somebody from the codegen team needs to sign-off on the JIT changes.

cc @JulieLeeMSFT

**Comment by JulieLeeMSFT** [comment]
@EgorBo, please do code review for Filip's compact unwinding PR.

**Review by kunalspathak — COMMENTED** [review]
I went over it and they sounds reasonable. @BruceForstall  - you might want to double check the changes in `lclvars.cpp`. I don't understand those much.

File: src/coreclr/jit/compiler.hpp [hunk 1]
```diff
@@ -2807,6 +2807,16 @@ inline
     {
         *pBaseReg = REG_SPBASE;
     }
+#elif defined(TARGET_ARM64)
+    if (FPbased && !codeGen->isFramePointerRequired() && varOffset < 0 &&
```
[the remark this answers is not in the document]
Comment (filipnavara) [reply src/coreclr/jit/compiler.hpp:2811]: The optimization is valid even for non-NativeAOT scenarios. The reason I explicitly opted not to guard it is that this is practically only reachable in stress tests that enable `SaveFpLrWithAllCalleeSavedRegisters` on CoreCLR and the general guidance was that we rely on CoreCLR stress testing coverage for JIT.

File: src/coreclr/jit/codegenarm64.cpp [hunk 2]
```diff
@@ -845,12 +845,19 @@ void CodeGen::genSaveCalleeSavedRegisterGroup(regMaskTP regsMask, int spDelta, i
 
     for (int i = 0; i < regStack.Height(); ++i)
     {
-        RegPair regPair = regStack.Bottom(i);
+        RegPair regPair = genReverseAndPairCalleeSavedRegisters ? regStack.Top(i) : regStack.Bottom(i);
```
[the remark this answers is not in the document]
Comment (filipnavara) [reply src/coreclr/jit/codegenarm64.cpp:848]: Not sure how an `#ifdef` could be used here. There's a single universal build of the ARM64 Unix JIT and it's the same for Crossgen2 and ILC.

File: src/coreclr/jit/compiler.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (filipnavara) [reply src/coreclr/jit/compiler.hpp:2811]: The changes in `gen[Restore/Save]CalleeSavedRegisterGroup` are guarded by `genReverseAndPairCalleeSavedRegisters` which is only set for NativeAOT/Apple. Additional guards would only contribute to the throughput cost. The change where LR/FP is separated into its own condition is no-op for other ARM64 scenarios, it just moves one iteration of the loop into its own condition. We don't want to reverse the order for them so this was the easiest way to achieve the desired behavior without too many `if`s.

File: src/coreclr/jit/codegenarm64.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kunalspathak) [reply src/coreclr/jit/codegenarm64.cpp:848]: I meant `compiler->IsTargetAbi(CORINFO_NATIVEAOT_ABI) ` check

File: src/coreclr/jit/compiler.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (kunalspathak) [reply src/coreclr/jit/compiler.hpp:2811]: that sounds good to me

File: src/coreclr/jit/codegenarm64.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (filipnavara) [reply src/coreclr/jit/codegenarm64.cpp:848]: I'd expect that to make things worse than checking a single [more local] boolean. If this is a concern we can try to manually duplicate the loop and move the check outside of the loop.

**Comment by xoofx** [comment]
@filipnavara I'm investigating how much trouble it would be to do a fast stack walking on macOS arm64 for profiling purpose, and I have a question. The [Writing ARM64 code for Apple platforms](https://developer.apple.com/documentation/xcode/writing-arm64-code-for-apple-platforms) mandates that:

> _The frame pointer register (x29) must always address a valid frame record. Some functions — such as leaf functions or tail calls — may opt not to create an entry in this list. As a result, stack traces are always meaningful, even without debug information._

So, while trying to reconnect the dots with the various issues related to this PR, I'm unclear what kind of frames the JIT today generates for Apple ARM64? Is it following the fp recommendation above?

**Comment by filipnavara** [comment]
> So, while trying to reconnect the dots with the various issues related to this PR, I'm unclear what kind of frames the JIT today generates for Apple ARM64? Is it following the fp recommendation above?

TL;DR: Yes, it is following the recommendation AFAICT. It doesn't follow the relative stack layout of FP/LR in relation to locals and callee saved registers. Unless you need to do precise unwinding you should be fine.

The long description is https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/jit/arm64-jit-frame-layout.md (not all frame types are used, eg. frame type 9).

**Comment by JulieLeeMSFT** [comment]
@EgorBo, and @kunalspathak, this PR is waiting for JIT sign off.

File: src/coreclr/jit/lclvars.cpp [hunk 3]
```diff
@@ -6859,6 +6852,44 @@ void Compiler::lvaAssignVirtualFrameOffsetsToLocals()
...
+    {
+        if (IsTargetAbi(CORINFO_NATIVEAOT_ABI) && TargetOS::IsApplePlatform &&
+            (!codeGen->isFramePointerRequired() || codeGen->genTotalFrameSize() < 0x100))
+        {
+            codeGen->SetSaveFpLrWithAllCalleeSavedRegisters(true);
+        }
+        else
+        {
+            // Default configuration
+            codeGen->SetSaveFpLrWithAllCalleeSavedRegisters((getNeedsGSSecurityCookie() && compLocallocUsed) ||
```
Comment (kunalspathak) [inline src/coreclr/jit/lclvars.cpp:6879]: shouldn't this be `(getNeedsGSSecurityCookie() || compLocallocUsed)`?

**Review by kunalspathak — CHANGES_REQUESTED** [review]
Can you please merge with main? I want to run some jitstress tests because `compJitSaveFpLrWithCalleeSavedRegisters` was updated.

File: src/coreclr/jit/lclvars.cpp (hunk 3 above)
Comment (filipnavara) [reply src/coreclr/jit/lclvars.cpp:6879]: That's just a moved code from [here](https://github.com/dotnet/runtime/blob/717bc30a0ed2a32f05f90e288eaaf7c02712c916/src/coreclr/jit/lclvars.cpp#L6059) with unchanged condition. I think it was intentional.

```
    // Decide where to save FP and LR registers. We store FP/LR registers at the bottom of the frame if there is
    // a frame pointer used (so we get positive offsets from the frame pointer to access locals), but not if we
    // need a GS cookie AND localloc is used, since we need the GS cookie to protect the saved return value,
    // and also the saved frame pointer. See CodeGen::genPushCalleeSavedRegisters() for more details about the
    // frame types. Since saving FP/LR at high addresses is a relatively rare case, force using it during stress.
    // (It should be legal to use these frame types for every frame).
```

**Comment by filipnavara** [comment]
Rebased over dotnet/main.

**Comment by kunalspathak** [comment]
There is one failure I see for both linux/windows arm64 in libraries-jitstress. Not sure if it is existing one, but can you please double check on your end if it is because of this change?

**Comment by filipnavara** [comment]
> There is one failure I see for both linux/windows arm64 in libraries-jitstress. Not sure if it is existing one, ...

Do week have weekly/daily runs of the JIT stress tests for reference? If not, would it be possible to run them on a dummy PR with no change?

I'll try to check it locally but it's going to take me a while to get the infrastructure running on my Windows machine.

**Comment by MichalStrehovsky** [comment]
> Do week have weekly/daily runs of the JIT stress tests for reference? If not, would it be possible to run them on a dummy PR with no change?

The runs should be publicly visible at https://dev.azure.com/dnceng-public/public/_build?definitionId=138&_a=summary. Looks like they are daily.

**Comment by kunalspathak** [comment]
I have retriggered the run to make sure that it is really caused by this PR. The latest run is green. https://dev.azure.com/dnceng-public/public/_build/results?buildId=910114&view=results

**Comment by filipnavara** [comment]
Thanks. It seems to be caused by the PR. I am trying to repro it locally on my Windows ARM machine so I can compare the JIT dumps.

**Comment by filipnavara** [comment]
I think the issue is specific to Tier1-OSR methods. My suspicion is that the optimization [here](https://github.com/dotnet/runtime/pull/107766/files#diff-30076431057df5c991b5e11659842a02ece8bb0669754fd1f42131272305bcc0R2811-R2820) calculates the wrong offset in that case. I'm not particularly interested in stressing that code path for OSR methods since they are not used on NativeAOT. I'll try to disable the optimization for OSR and see if it helps.

**Comment by filipnavara** [comment]
I re-run the tests locally and they no longer fail. Can you retrigger the test runs please?

**Comment by kunalspathak** [comment]
@filipnavara - the original failures no longer repros and some tests are failing which is most likely https://github.com/dotnet/runtime/issues/83659, but can you please double check if jitstress is kicking in for it?

**Comment by filipnavara** [comment]
> ...but can you please double check if jitstress is kicking in for it?

Yes, I put breakpoints on the code path in WinDBG and they do get hit during the stress test run:

image

**Comment by kunalspathak** [comment]
So they tests are failing with the new jitstress switch?

**Comment by filipnavara** [comment]
> So they tests are failing with the new jitstress switch?

No, they are NOT failing, and they stress the new code path.

**Review by kunalspathak — APPROVED** [review]
LGTM

**Comment by EgorBo** [comment]
@filipnavara Looks like this PR introduced GC hole like asserts in gcstress pipeline (linux-arm64), I've just confirmed it.

The simplest way to reproduce is to take [BufferMemmove](https://github.com/dotnet/runtime/issues/111388) test and run with
```
set DOTNET_TieredCompilation=0
set DOTNET_GCStress=0xC
set DOTNET_JitStress=2
set DOTNET_ReadyToRun=0
```
(many times)

cc @dotnet/jit-contrib
